src/sinks/socket.rs

```diff
@@ -155,4 +155,106 @@ mod test {
+
+    // This is a test that checks that we properly receieve all events in the
+    // case of a proper server side write side shutdown.
+    //
+    // This test basically sends 10 events shutsdown the server and forces a
+    // reconnect. It then forces another 10 events through and we should get a
+    // total of 20 events.
+    //
+    // If this test hangs that means somewhere we are not collecting the correct
+    // events.
+    #[test]
+    fn tcp_stream_detects_disconnect() {
+        let addr = next_addr();
+        let config = SocketSinkConfig {
+            mode: Mode::Tcp(TcpSinkConfig {
+                address: addr.to_string(),
+                encoding: Encoding::Text.into(),
+                // TODO: enable TLS here since this is where
+                // ran into the issue before!
+                tls: None,
+            }),
+        };
+        let mut rt = runtime();
+        let context = SinkContext::new_test(rt.executor());
+        let (sink, _healthcheck) = config.build(context).unwrap();
+
+        use std::sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc,
+        };
+
+        let close = Arc::new(tokio::sync::Notify::new());
+        let counter = Arc::new(AtomicUsize::new(0));
+
+        let close1 = close.clone();
+        let counter1 = counter.clone();
+        let jh = rt.spawn_handle(async move {
+            use tokio::{io::AsyncReadExt, net::TcpListener};
+
+            let mut listener = TcpListener::bind(&addr).await.unwrap();
+
+            // Only accept two connections after the second connection is done
+            // we can exit.
+            for _ in 0..2 {
+                let (mut conn, _) = listener.accept().await.unwrap();
+
+                loop {
+                    let mut buf = vec![0u8; 11];
+
+                    tokio::select! {
+                        _ = close1.notified() => {
+                            conn.shutdown(std::net::Shutdown::Write).unwrap();
+                            break;
+                        }
+
+                        res = conn.read(&mut buf) => {
+                            let n = res.unwrap();
+
+                           if  n == 0 {
+                               break;
+                           } else {
+                               counter1.fetch_add(1, Ordering::Relaxed);
+                           }
+                        }
+                    };
+                }
+            }
+        });
+
+        let (_, events) = random_lines_with_stream(10, 10);
+        let pump = sink.send_all(events);
+        let (sink, _) = rt.block_on(pump).unwrap();
+
+        // Loop and check for 10 events, we should always get 10 events. Once,
+        // we have 10 events we can tell the server to shutdown to simulate the
+        // remote shutting down on an idle connection.
+        for _ in 0..100 {
+            let amnt = counter.load(Ordering::SeqCst);
+
+            if amnt == 10 {
+                close.notify();
```

Comment:
Couldn't we just have the server count to ten and shutdown itself?

Reply:
Yes, but I really liked the conceptual idea to shutdown the server outside of a count. Aka driven from outside. I'd prefer to keep it this way, if we find its bad in a refactor we can change it then.

Reply:
I'm just curious why. In the real world, it is the server's decision when to shut down the connection and doing it here based on the count seems in line with that? And it also addresses the problem of sleeping to ensure the server had read all ten.

Reply:
If we don't have this notify/count how would the client know when to send the next 10 events? Does it sleep for a period before pumping new events?